### ISSUE TITLE
Full indent for list sub-items

### ISSUE THREAD

**Issue by ivanpirog** [pbek/QOwnNotes#1581]
### Actual behaviour:

![QON Actual](https://user-images.githubusercontent.com/8699212/72341918-292cf400-36d4-11ea-8fd9-e2bc63d795b3.png)

### Expected behaviour:

![QON Expected](https://user-images.githubusercontent.com/8699212/72341924-2c27e480-36d4-11ea-9fe0-e9a353a8fc3e.png)

**Comment by pbek** [comment]
Can you please explain in more detail and step by step what you did, what happened and what you expected.

**Comment by ivanpirog** [comment]
Ok. I press Tab to indent the new list item. The Tab key indents 4 spaces from the beginning of a line. Then I write a long sentence that wraps on the next line...

### Current behavior:

The wrapped text of a long subitem sentence is displayed from the very beginning of the line.

![QON Actual](https://user-images.githubusercontent.com/8699212/72343732-3815a580-36d8-11ea-804a-6125edd6c160.png)

The screenshot shows that the second line of the long subitem sentence starts from the very beginning of the line. The arrows indicate this.

### Expected Behavior:

The wrapped text of a long subitem sentence is displayed indented, which is equal to the number of spaces from the beginning of the line, including the asterisk and a space after it.

![QON Expected](https://user-images.githubusercontent.com/8699212/72343844-83c84f00-36d8-11ea-8d22-dbac00247545.png)

The screenshot shows that the second line of the long subitem sentence does not start at the very beginning of the line, but indented.

**Comment by pbek** [comment]
Thank you for your explanation. I don't know if you can even alter the soft wrap mechanism and if it is possible it sounds like a lot of work. :grimacing: 

@Waqar144, have you ever played with indents of the soft-wrapping?

**Comment by ivanpirog** [comment]
> if it is possible it sounds like a lot of work. 😬

Yeah, I already looked at the source code and realized that the task is difficult (at first glance).

**Comment by ivanpirog** [comment]
I look at:

void QTextBlockFormat::setIndent ( int indentation )
void QTextCursor::setBlockFormat ( const QTextBlockFormat & format )

But it looks like it sets the indent for the entire paragraph.

If it is possible to set both indents for the entire paragraph and indent from the beginning of the line, then indent from the beginning of the line can be made negative.

**Comment by ivanpirog** [comment]
Possible steps:

1. Get a list item string
2. Get spaces from the start of the line: ^(\s+)
3. Get paragraph font size
4. SpacesWidthInPixels = Get width in pixels of spaces from the start of the line using QFontMetrics::width (https://doc.qt.io/archives/qt-4.8/qfontmetrics.html#width)
5. QTextBlockFormat::setTextIndent(-SpacesWidthInPixels) 
6. QTextBlockFormat::setLeftMargin(SpacesWidthInPixels)

**Comment by Waqar144** [comment]
> Thank you for your explanation. I don't know if you can even alter the soft wrap mechanism and if it is possible it sounds like a lot of work. grimacing
> 
> @Waqar144, have you ever played with indents of the soft-wrapping?

Nah. I always understoop wrapping as something only for the end of the line, not the start of the next one :laughing:

**Comment by Waqar144** [comment]
> Possible steps:
> 
>     1. Get a list item string
> 
>     2. Get spaces from the start of the line: ^(\s+)
> 
>     3. Get paragraph font size
> 
>     4. SpacesWidthInPixels = Get width in pixels of spaces from the start of the line using QFontMetrics::width (https://doc.qt.io/archives/qt-4.8/qfontmetrics.html#width)
> 
>     5. QTextBlockFormat::setTextIndent(-SpacesWidthInPixels)
> 
>     6. QTextBlockFormat::setLeftMargin(SpacesWidthInPixels)

@ivanpirog looks like you are already on way to solving the issue :rocket:

**Comment by pbek** [comment]
Yes, https://doc.qt.io/qt-5/qtextblockformat.html#setTextIndent and https://doc.qt.io/qt-5/qtextblockformat.html#setLeftMargin sound like a good way.

**Comment by pbek** [comment]
I tried some stuff, but had my issues with it...

```cpp
QTextBlockFormat blockFormat = currentBlock().blockFormat();
// didn't do anything for me
blockFormat.setTextIndent(100);
// didn't do anything for me
blockFormat.setLeftMargin(50);

auto* myCursor = new QTextCursor(currentBlock());
// this prevents "undo" 
myCursor->setBlockFormat(blockFormat);
```

Issues like broken "undo" already did came up previously when I tried to introduce a bottom margin for headlines.

**Comment by ivanpirog** [comment]
> ```c++
> QTextBlockFormat blockFormat = currentBlock().blockFormat();
> // didn't do anything for me
> blockFormat.setTextIndent(100);
> // didn't do anything for me
> blockFormat.setLeftMargin(50);
> 
>

Yeah, I have the same results :(

**Comment by Waqar144** [comment]
This bug is a **won't fix** because `QMarkdownTextEdit` is pure plaintext and doesn't support any complicated formatting.

**Comment by pbek** [comment]
Which bug? The undo?

**Comment by Waqar144** [comment]
The indenting one. (Bug/feature request)

'Undo' problem isn't a bug according to my understanding.

**Comment by pbek** [comment]
If a QPlainTextEdit can't do it it currently can't be done. Except we want to pay the price for switching back to a QTextEdit (speed, memory consumption, possibly missing features that currently work)

**Comment by nonplayer** [comment]
I know this is way over a year old, but hey it's still technically open. I came here to post a request for this same feature. I don't know if it would be any easier at all to implement now, but I figured I'd ask. Bullet-aligned list-wrapping indents would be wonderful. Thanks.

**Comment by pbek** [comment]
So far there is no solution for it.

**Comment by hbarel** [comment]
I can only add a "+1" for the need. I was just about to post it as a suggested feature and saw that it's already there...

I realize it is not doable, though.

**Comment by carhe** [comment]
Same for me, I was on the way to suggest the feature and found this issue.

For people structuring (longer) thougths in hierarchical lists the indent feature is quite important. Without indentation the visual structure of the text gets lost, and with nested items it gets unclear which of the items the unintended text continues. The current workaround is to indent the text manually, which is really awkward because changing the window size and in consequence the line length breaks everything (this typically happens when one wants to export or print the text).

Just some background to support my +1.

**Comment by pbek** [comment]
## 26.2.14

- Added hanging indentation for wrapped Markdown list lines in the note text edit so that
  continuation lines align with the list content rather than the list marker
  (for [#1581](https://github.com/pbek/QOwnNotes/issues/1581))

**Comment by pbek** [comment]
## 26.3.1

- Added a setting to toggle hanging indentation for wrapped Markdown list items
  in the note text edit (for [#1581](https://github.com/pbek/QOwnNotes/issues/1581))

**Comment by nonplayer** [comment]
@pbek Thanks for adding this option! One last suggestion: As this option directly affects indents, it should probably be moved to the specific "Tabs and Indents" sub-section of the settings, instead of the more generalized "Editor Options" sub-section.

**Comment by pbek** [comment]
Good point, thank you for mentioning!

**Comment by pbek** [comment]
## 26.3.2

- Moved the setting for hanging indentation in wrapped Markdown list items to the
  Tabs and Indents panel in the settings dialog (for [#1581](https://github.com/pbek/QOwnNotes/issues/1581))

**Comment by pbek** [comment]
There now is a new release, could you please test it and report if it works for you?

**Comment by nonplayer** [comment]
@pbek Just updated, and it looks good. Works exactly as desired. Thanks for this, my notes now look so much better.

**Comment by pbek** [comment]
Great, thank you for testing!
